File: Example/OpAdxDemo_iOS/OpAdxDemo_iOS/Classes/BannerViewController.m
//
//  BannerViewController.m
//  OpAdxDemo_iOS
//
//  Created by Luan Chen on 2025/12/10.
//

#import "BannerViewController.h"
#import <OpAdxSdk/OpAdxSdk-Swift.h>


@interface BannerViewController()<OpAdxBannerAdDelegate>

@property (nonatomic, strong, nullable) OpAdxBannerAdBridge *bannerAdView;

@end

@implementation BannerViewController

- (BOOL)hasVideo {
    return YES;
}

- (NSString *)adFormatString {
    return @"Banner Ad";
}

- (void)loadAd {
    [self destroyAd];
    
    __weak typeof(self) weakSelf = self;
    NSString *placementID = self.placementId;
    
    OpAdxBannerAdBridge *bannerAd = [[OpAdxBannerAdBridge alloc] initWithPlacementId:placementID adSize:OpAdxAdSize.BANNER_MREC];
    self.bannerAdView = bannerAd;
    bannerAd.delegate = self;
    [bannerAd loadAd];
}

- (void)showAd {
    if (!self.bannerAdView) return;

#ifdef DEBUG
    // Debug 模式下不检查 isAdInvalidated
#else
    if (self.bannerAdView.isAdInvalidated) {
        [self.logView print:@"Ad is invalidated"];
        [self destroyAd];
        return;
    }
#endif
    
    UIView *bannerAdView = self.bannerAdView;
    bannerAdView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.adContainer addSubview:bannerAdView];
    NSArray<NSLayoutConstraint *> *constraints = @[
        // 图标
        [bannerAdView.centerXAnchor constraintEqualToAnchor:self.adContainer.centerXAnchor],
        [bannerAdView.centerYAnchor constraintEqualToAnchor:self.adContainer.centerYAnchor],
        [bannerAdView.widthAnchor constraintEqualToAnchor:self.adContainer.widthAnchor],
        [bannerAdView.heightAnchor constraintEqualToAnchor:self.adContainer.heightAnchor],
//        [bannerAdView.widthAnchor constraintEqualToConstant:OpAdxAdSize.BANNER_MREC.width],
//        [bannerAdView.heightAnchor constraintEqualToConstant:OpAdxAdSize.BANNER_MREC.height],
    ];
    
    [NSLayoutConstraint activateConstraints:constraints];
    
    [self enableDestroyAd];
    [self disableShowAd];
}

- (void)destroyAd {
    [super destroyAd];
    [self.bannerAdView destroy];
    self.bannerAdView = nil;
    
    [self disableShowAd];
    [self disableDestroyAd];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.bannerAdView pause];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.bannerAdView resume];
}

#pragma mark - OpAdxBannerAdDelegate

- (void)bannerAdDidLoad:(OpAdxBannerAdBridge *)bannerAd {
    NSLog(@"[ADX] Banner广告加载成功: %@", bannerAd.placementID);
    
    [self.logView print:[NSString stringWithFormat:@"Loaded, adSize: %@ ", bannerAd.adSize]];
   
    [self enableShowAd];
    [self enableDestroyAd];
    
}

- (void)bannerAd:(OpAdxBannerAdBridge *)bannerAd didFailWithError:(OpAdxAdError *)error {
    NSLog(@"[ADX] Banner广告加载失败: %@", error);
    [self.logView print:error.message];
    
}

- (void)bannerAdDidClick:(OpAdxBannerAdBridge *)bannerAd {
    NSLog(@"[ADX] Banner广告被点击: %@", bannerAd.placementID);
    [self.logView print:@"onAdClicked"];
}

- (void)bannerAdWillLogImpression:(OpAdxBannerAdBridge *)bannerAd {
    NSLog(@"[ADX] Banner广告展示: %@", bannerAd.placementID);
    [self.logView print:@"onAdImpression"];
    
}

- (void)dealloc {
    [self.bannerAdView destroy];
    self.bannerAdView = nil;
}

@end

File: Example/OpAdxDemo_Swift/OpAdxSdkDemo/BannerViewController.swift
//
//  BannerViewController.swift
//  OperaAdxSDK
//
//  Created by Luan Chen on 2025/11/4.
//

import UIKit
import OpAdxSdk

class BannerViewController: BaseViewController {
    
    private var bannerAdView: OpAdxBannerAdView?
    
    override var hasVideo: Bool {
        return true
    }
    
    override var adFormatString: String {
        return "Banner Ad"
    }
    
    override func loadAd() {
        destroyAd()
        
        if bannerAdView == nil {
            bannerAdView = OpAdxBannerAdView()
        }
        
        logView.print("Loading ...")
        
        bannerAdView?.setPlacementId(placementId!)
        bannerAdView?.setAdSize(.BANNER_MREC)

        // 创建遵循 BannerAdListener 协议的实例
        let listener = OpAdxBannerAdListenerImp(
           onAdLoaded: { [weak self] bannerAdInfo in
               guard let self = self, let bannerAdInfo = bannerAdInfo as? OpAdxBannerAdInfo else { return }
               self.logView.print("Loaded \(bannerAdInfo.adSize), refreshCount: \(bannerAdInfo.refreshCount)")
               self.enableShowAd()
               self.enableDestroyAd()
           },
           onAdFailedToLoad: { [weak self] error in
               guard let self = self else { return }
               self.logView.print(error.message)
           },
           onAdImpression: { [weak self] in
               guard let self = self else { return }
               self.logView.print("Impression!")
           },
           onAdClicked: { [weak self] in
               guard let self = self else { return }
               self.logView.print("Clicked!")
           }
        )
        bannerAdView?.loadAd(listener: listener)
    }
    
    override func showAd() {
        guard let bannerAdView = bannerAdView else { return }
#if DEBUG
#else
        if bannerAdView.isAdInvalidated {
            logView.print("Ad is invalidated")
            destroyAd()
            return
        }
#endif
        
        bannerAdView.translatesAutoresizingMaskIntoConstraints = false
        adContainer.addSubview(bannerAdView)
        
        NSLayoutConstraint.activate([
            bannerAdView.centerXAnchor.constraint(equalTo: adContainer.centerXAnchor),
            bannerAdView.centerYAnchor.constraint(equalTo: adContainer.centerYAnchor),
            bannerAdView.widthAnchor.constraint(equalToConstant: 300),
            bannerAdView.heightAnchor.constraint(equalToConstant: 250)
        ])
        
        enableDestroyAd()
        disableShowAd()
    }
    
    override func destroyAd() {
        super.destroyAd()
        bannerAdView?.destroy()
        bannerAdView = nil
        
        adContainer.subviews.forEach { $0.removeFromSuperview() }
        disableShowAd()
        disableDestroyAd()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        bannerAdView?.pause()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        bannerAdView?.resume()
    }
    
    deinit {
        bannerAdView?.destroy()
        bannerAdView = nil
    }
}

